contracts/EtomicSwapV2.sol

```diff
@@ -0,0 +1,476 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract EtomicSwapV2 {
+    using SafeERC20 for IERC20;
+
+    enum MakerPaymentState {
+        Uninitialized,
+        PaymentSent,
+        TakerSpent,
+        MakerRefunded
+    }
+
+    struct MakerPayment {
+        bytes20 paymentHash;
+        uint32 paymentLockTime;
+        MakerPaymentState state;
+    }
+
+    event MakerPaymentSent(bytes32 id);
+    event MakerPaymentSpent(bytes32 id, bytes32 secret);
+    event MakerPaymentRefundedTimelock(bytes32 id);
+    event MakerPaymentRefundedSecret(bytes32 id);
+
+    mapping(bytes32 => MakerPayment) public makerPayments;
+
+    enum TakerPaymentState {
+        Uninitialized,
+        PaymentSent,
+        TakerApproved,
+        MakerSpent,
+        TakerRefunded
+    }
+
+    struct TakerPayment {
+        bytes20 paymentHash;
+        uint32 immediateRefundTime;
```

Comment:
I would like to discuss it just in case because `TakerApproved` option requires more gas for swap. Will post more info a bit later :slightly_smiling_face:

Reply:
I had to change `immediateRefundTime` to `preApproveLockTime` to fix another note, so this is no longer actual. If we try to introduce additional `immediateRefundTime`, TakerPayment struct size will be over 32 bytes, increasing gas consumption of the payment. So, having `takerApprove` seems the only way to go for now :slightly_smiling_face: